beacon-chain/rpc/validator/proposer.go

```diff
@@ -194,6 +206,128 @@ func (vs *Server) eth1Data(ctx context.Context, slot uint64) (*ethpb.Eth1Data, e
 	return eth1Data, nil
 }
 
+// eth1DataMajorityVote determines the appropriate eth1data for a block proposal using an extended
+// simple voting algorithm - voting with the majority. The algorithm for this method is as follows:
+//  - Determine if we are in the "head" of the voting period. If yes, use the simple voting algorithm for proposal.
+//  - Determine the timestamp for the start slot for the current eth1 voting period.
+//  - Determine the timestamp for the start slot for the previous eth1 voting period.
+//  - Determine the most recent eth1 block before each timestamp.
+//  - Subtract the current period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting upper bound.
+//  - Subtract the previous period's eth1block.number by ETH1_FOLLOW_DISTANCE to determine the voting lower bound.
+//  - Filter out votes on unknown blocks and blocks which are outside of the range determined by the lower and upper bounds.
+//  - If no blocks are left after filtering, use the current period's most recent eth1 block for proposal.
+//  - Determine the vote with the highest count. Prefer the vote with the highest eth1 block height in the event of a tie.
+//  - This vote's block is the eth1block to use for the block proposal.
+// TODO: "Genesis" period and previous period
+// TODO: Feature flag
+func (vs *Server) eth1DataMajorityVote(ctx context.Context, slot uint64) (*ethpb.Eth1Data, error) {
+	ctx, cancel := context.WithTimeout(ctx, eth1dataTimeout)
+	defer cancel()
+
+	if vs.MockEth1Votes {
+		return vs.mockETH1DataVote(ctx, slot)
+	}
+	if !vs.Eth1InfoFetcher.IsConnectedToETH1() {
+		return vs.randomETH1DataVote(ctx)
+	}
+	eth1DataNotification = false
+
+	headOfVotingPeriod :=
+		slot%(params.BeaconConfig().EpochsPerEth1VotingPeriod*params.BeaconConfig().SlotsPerEpoch) <
+			params.BeaconConfig().HeadOfVotingPeriodLength
+	if headOfVotingPeriod {
+		// TODO: Is it OK that `context.WithTimeout` will be invoked again?
+		return vs.eth1Data(ctx, slot)
```

Comment:
I am invoking `eth1Data()` from within `eth1DataMajorityVote()` and both functions begin with `context.WithTimeout(ctx, eth1dataTimeout)`. Is it a problem that the context function will be invoked twice?

Reply:
yes that is fine